### ISSUE TITLE
llInjectorError: No provider for RadioControlRegistry!  

### ISSUE THREAD

**Issue by zip-fa** [angular/angular#50312]
### Which @angular/* package(s) are the source of the bug?

forms

### Is this a regression?

Yes

### Description

Not sure if it's related to @angular/components or @angular/forms

After upgrading from v14 to v16 input type="radio" inside CDK dialog throws error.
Other form inputs work as expected - only radio type affected.
When i render my dialog's component from template directly it works as expected.

### Please provide a link to a minimal reproduction of the bug

https://stackblitz.com/edit/angular-f2m372?file=src/test.dialog.ts

### Please provide the exception or error you saw

```true
Error: Uncaught (in promise): NullInjectorError: R3InjectorError(Environment Injector)[RadioControlRegistry -> RadioControlRegistry]: 
  NullInjectorError: No provider for RadioControlRegistry!
NullInjectorError: R3InjectorError(Environment Injector)[RadioControlRegistry -> RadioControlRegistry]: 
  NullInjectorError: No provider for RadioControlRegistry!
    at NullInjector.get (core.mjs:8712:27)
    at R3Injector.get (core.mjs:9138:33)
    at R3Injector.get (core.mjs:9138:33)
    at ChainedInjector.get (core.mjs:12938:36)
    at lookupTokenUsingModuleInjector (core.mjs:4388:39)
    at getOrCreateInjectable (core.mjs:4433:12)
    at Module.ɵɵdirectiveInject (core.mjs:10704:12)
    at NodeInjectorFactory.RadioControlValueAccessor_Factory [as factory] (forms.mjs:68:84)
    at getNodeInjectable (core.mjs:4618:44)
    at instantiateAllDirectives (core.mjs:11524:27)
    at resolvePromise (zone.js:1193:31)
    at resolvePromise (zone.js:1147:17)
    at zone.js:1100:17
    at zone.js:1116:33
    at asyncGeneratorStep (asyncToGenerator.js:10:5)
    at _next (asyncToGenerator.js:22:9)
    at asyncToGenerator.js:27:7
    at new ZoneAwarePromise (zone.js:1411:21)
    at asyncToGenerator.js:19:12
    at Dialog.open (dialog.ts:33:8)
```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 16.0.1
Node: 16.14.2
Package Manager: npm 8.5.0
OS: win32 x64

Angular: 16.0.1
... animations, cdk, cli, common, compiler, compiler-cli, core
... forms, localize, material, platform-browser
... platform-browser-dynamic, platform-server, router
... service-worker

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1600.1
@angular-devkit/build-angular   16.0.1
@angular-devkit/core            16.0.1
@angular-devkit/schematics      16.0.1
@nguniversal/builders           16.0.1
@nguniversal/express-engine     16.0.1
@schematics/angular             16.0.1
rxjs                            7.8.1
typescript                      5.0.4
```

**Comment by JeanMeche** [comment]
Hi I wasn't able to reproduce the error with information you gave us. Could try write a minimal reproduction on stackblitz ? Thx !

**Comment by zip-fa** [comment]
> Hi I wasn't able to reproduce the error with information you gave us. Could try write a minimal reproduction on stackblitz ? Thx !

Updated my issue. You could try it

**Comment by JeanMeche** [comment]
`FormsModule` must be provided at the root of your app.

**Comment by zip-fa** [comment]
But why it was working before v16?:)

**Comment by zip-fa** [comment]
And you see dialog without bug works fine

**Comment by zip-fa** [comment]
Also, this will work:

https://stackblitz.com/edit/angular-mjkb7d?file=src%2Fmain.ts,src%2Ftest.module.ts,src%2Ftest.dialog.ts,src%2Ftest2.dialog.ts

**Comment by JeanMeche** [comment]
The code with the modal doesn't work with 14/15. 

Your issue lie with the usage of the CDK.  The `DialogModule` doesn't have access to the providers imported by the App component. 

1. In your first example, it the injector of the CDK module that is resolving the provider.
2. In your second example it's the component that is resolving the provider and that works thanks to the import.

**Comment by JeanMeche** [comment]
It's a short comming of the implementation of `RadioControlRegistry`, it's not `providedIn:'root'`. 

https://github.com/angular/angular/blob/a09f88934d7b1380d3a6f08c64e71901686a18cc/packages/forms/src/directives/radio_control_value_accessor.ts#L44

There is a PR to fix that #50037

**Comment by JoostK** [comment]
This isn't working for the same reason as https://github.com/angular/angular/issues/47448#issuecomment-1268830214; the `Dialog` service is provided in the app injector, whereas `ReactiveFormsModule` is only imported by the standalone component `App`. The dialog components are opened from within the context of `Dialog` in root, so the `App`'s providers are not in scope (they are children of the root).

It starts to work once you add `importProvidersFrom(ReactiveFormsModule)` to the `bootstrapApplication` call.

Although confusing, this is working as designed. The change in #50037 alleviates this particular instance.
